### PR TITLE
Fix for Homogenous precision #14242

### LINKED ISSUE #14242: Homogenous precision
Can we have an `inf` at https://github.com/opencv/opencv/blob/master/modules/calib3d/src/fundam.cpp#L970 if  `1e-39>=sptr[i].z<=1e-45`?

### PR DESCRIPTION
**Pull request by ostarling**
resolves #14242

### CODE REVIEW

File: modules/calib3d/src/fundam.cpp [hunk 1]
```diff
@@ -944,7 +944,7 @@ void cv::convertPointsFromHomogeneous( InputArray _src, OutputArray _dst )
             Point2f* dptr = dst.ptr<Point2f>();
             for( i = 0; i < npoints; i++ )
             {
-                float scale = sptr[i].z != 0 ? 1.f/sptr[i].z : 1.f;
+                float scale = (std::fabs(sptr[i].z) > FLT_EPSILON) ? 1.f/sptr[i].z : 1.f;
```
Comment (csukuangfj) [inline modules/calib3d/src/fundam.cpp]: I would suggest replacing
`FLT_EPSILON`
with
`std::numeric_limits<float>::epsilon()`.

**Comment by csukuangfj** [comment]
duplicate #14289

**Comment by ostarling** [comment]
Sure, I’ll do that. Do you have anything against turning these 6 or so similar changes to an inline function?

File: modules/calib3d/src/fundam.cpp (hunk 1 above)
Comment (ostarling) [reply modules/calib3d/src/fundam.cpp]: Changed as suggested

File: modules/calib3d/src/fundam.cpp [hunk 2]
```diff
@@ -944,7 +948,7 @@ void cv::convertPointsFromHomogeneous( InputArray _src, OutputArray _dst )
             Point2f* dptr = dst.ptr<Point2f>();
             for( i = 0; i < npoints; i++ )
             {
-                float scale = sptr[i].z != 0 ? 1.f/sptr[i].z : 1.f;
```
Comment (mshabunin) [inline modules/calib3d/src/fundam.cpp]: `sptr[i]` is `int` here, we can compare it with zero safely.

File: modules/calib3d/src/fundam.cpp [hunk 3]
```diff
@@ -954,7 +958,7 @@ void cv::convertPointsFromHomogeneous( InputArray _src, OutputArray _dst )
             Point3f* dptr = dst.ptr<Point3f>();
             for( i = 0; i < npoints; i++ )
             {
-                float scale = sptr[i][3] != 0 ? 1.f/sptr[i][3] : 1.f;
```
Comment (mshabunin) [inline modules/calib3d/src/fundam.cpp]: `sptr[i]` is `int` here, we can compare it with zero safely.

File: modules/calib3d/src/fundam.cpp (hunk 2 above)
Comment (thangktran) [reply modules/calib3d/src/fundam.cpp]: @mshabunin i believe `sptr[i]` is `float` from here
https://github.com/opencv/opencv/blob/master/modules/calib3d/src/fundam.cpp#L966

Comment (ostarling) [reply modules/calib3d/src/fundam.cpp]: Oh, yes, sorry. Will revert this and the other similar line back.

File: modules/calib3d/src/fundam.cpp [hunk 4]
```diff
@@ -909,6 +909,10 @@ void cv::computeCorrespondEpilines( InputArray _points, int whichImage,
     }
 }
 
+inline float scaleFor(float x){
```
Comment (alalek) [inline modules/calib3d/src/fundam.cpp]: "static" should be used for local functions.

File: modules/calib3d/src/fundam.cpp (hunk 2 above)
Comment (ostarling) [reply modules/calib3d/src/fundam.cpp]: reverted mistakenly changed lines where the value type was integer

File: modules/calib3d/src/fundam.cpp (hunk 4 above)
Comment (ostarling) [reply modules/calib3d/src/fundam.cpp]: Made the function static

File: modules/calib3d/src/fundam.cpp [hunk 5]
```diff
@@ -1000,7 +1004,7 @@ void cv::convertPointsFromHomogeneous( InputArray _src, OutputArray _dst )
             Point3d* dptr = dst.ptr<Point3d>();
             for( i = 0; i < npoints; i++ )
             {
-                double scale = sptr[i][3] != 0.f ? 1./sptr[i][3] : 1.;
+                double scale = scaleFor(sptr[i][3]);
```
Comment (thangktran) [inline modules/calib3d/src/fundam.cpp]: build failed due to conversion from `double` (`sptr[i][3]`) to `float` (parameter of `scaleFor()`).

Comment (ostarling) [reply modules/calib3d/src/fundam.cpp]: added an overloaded version of scaleFor for double

File: modules/calib3d/src/fundam.cpp [hunk 6]
```diff
@@ -909,6 +909,14 @@ void cv::computeCorrespondEpilines( InputArray _points, int whichImage,
     }
 }
 
+static inline double scaleFor(double x){
+    return (std::fabs(x) > std::numeric_limits<double>::epsilon()) ? 1./x : 1.;
```
Comment (alalek) [inline modules/calib3d/src/fundam.cpp]: Double computations are usually truncated on float epsilon value too.
These "double" results are used for intermediate calculations only (for better precision).
But finally, they are converted into floats in functions which works with images (like `wapr()`).

Comment (ostarling) [reply modules/calib3d/src/fundam.cpp]: changed to <float>::epsilon()

**Comment by bhack** [comment]
@alalek Sorry for cross-repository. What do you think about related issue in this PR using directly `1` as a scale factor. See https://github.com/arraiyopensource/kornia/pull/148#discussion_r288283863

**Comment by bhack** [comment]
@alalek Have we a problem also in `cvProjectPoints2Internal`? https://github.com/opencv/opencv/blob/master/modules/calib3d/src/calibration.cpp#L786-L787
How are you clipping not visible (in the camera plane) points?

**Comment by alalek** [comment]
Probably this code should be changed too.

**Comment by bhack** [comment]
The new issue is at https://github.com/opencv/opencv/issues/14703

**Pull request merged**